email/reporter.go

```diff
@@ -0,0 +1,551 @@
+func (r *Reporter) fileReport(report database.NCMECReport) error {
+	// convenience variables
+	logger := r.staticLogger
+	abuseDB := r.staticAbuseDatabase
+
+	// acquire a lock on the report
+	lock := abuseDB.NewReportLock(report.ID.Hex())
+	err := lock.Lock()
+	if err != nil {
+		return errors.AddContext(err, "could not acquire lock")
+	}
+
+	// defer the unlock
+	defer func() {
+		unlockErr := lock.Unlock()
+		if unlockErr != nil {
+			err = errors.Compose(err, errors.AddContext(unlockErr, "could not release lock"))
+			return
+		}
+	}()
+
+	// under lock, check whether the report has not been filed yet by another
+	// process, if so we simply return
+	current, err := abuseDB.FindReport(report.ID)
+	if err != nil {
+		return errors.AddContext(err, "could not find report")
+	}
+	if current.Filed {
+		return nil
+	}
+
+	// if the report has an NCMEC report id, it means something went wrong when
+	// trying to finish the report last time, so we try again.
+	if report.ReportID != 0 {
+		return r.finishReport(report)
+	}
+
+	// otherwise open a report with NCMEC
+	reportId, err := r.openReport(report)
+	if err != nil {
+		logger.Errorf("failed to open report, error '%v'", err)
+	}
+
+	// only if the report id is 0 we want to escape, if it is not 0 it means
+	// a report has been opened but something went wrong after opening the
+	// report, in that case we want to try and continue to finish the report
+	if reportId == 0 {
+		return err
+	}
+
+	// set the report id and finish the report
+	report.ReportID = reportId
+	err = r.finishReport(report)
+	if err != nil {
+		logger.Errorf("failed to finish report, error '%v'", err)
+		return err
+	}
+
+	return nil
+}
+
+// finishReport will finish the report with NCMEC
+func (r *Reporter) finishReport(report database.NCMECReport) error {
+	// convenience variables
+	logger := r.staticLogger
+
+	// finish the report with NCMEC
+	var reportErr string
+	res, err := r.staticClient.finishReport(report.ReportID)
+	if err == nil && res.ResponseCode != ncmecStatusOK {
+		err = fmt.Errorf("unexpected response code %v when finishing report '%v'", res.ResponseCode, report.ID.Hex())
+	}
+	if err != nil {
+		reportErr = err.Error()
+		logger.Errorf("failed to finish report %v, err '%v'", report.ReportID, err)
+	}
+
+	// update the email and set the report err and reported flag
+	err = r.staticAbuseDatabase.UpdateReportNoLock(report, bson.M{
+		"$set": bson.M{
+			"filed":     err == nil,
+			"filed_at":  time.Now().UTC(),
+			"filed_err": reportErr,
+
+			"report_id": report.ReportID,
+		},
+	})
+	if err != nil {
+		logger.Errorf("failed to update report %v, err '%v'", report.ID, err)
+		return err
+	}
+	return nil
+}
+
+// openReport will open a report with NCMEC for the given email, it will
+// decorate the abuse email with the report id
+func (r *Reporter) openReport(entity database.NCMECReport) (uint64, error) {
+	// convenience variables
+	logger := r.staticLogger
+
+	// unmarshal the report
+	var report report
```

Comment:
nit: Shadowing types with local variable names is not great. It doesn't cause any direct problems here or in all the other cases throughout the PR but I wanted to mention it.

Reply:
I updated it in one place where I could easily update it to `r`. It's a little bit annoying the module is called `Reporter` so I can't go with `r` here. I prefer `report report` over some weird looking variable name in this case. Changed it in the other place though.